gap/lqueue.gi

```diff
@@ -1,174 +1,224 @@
+            PlistQueueExpand(queue);
+        fi;
     fi;
-  elif tail + 1 <> head then
-    head := head - 1;
-    queue![QDATA][head] := el;
-    queue![QHEAD] := head;
-  else
-    PlistQueueExpand(queue);
-    head := Size(queue);
-    queue![QDATA][head] := el;
-    queue![QHEAD] := head;
-  fi;
 end);
 
 InstallGlobalFunction(PlistQueuePopFront,
 function(queue)
-  local head, tail, last, result;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  if head <> tail then
-    if head = last then
-      head := 1;
-      result := queue![QDATA][last];
-      queue![QDATA][last] := fail;
-    else
-      head := head + 1;
-      result := queue![QDATA][head-1];
-      queue![QDATA][head-1] := fail;
+    local head, tail, last, result;
+
+    if IsEmpty(queue) then
+        return fail;
     fi;
-    queue![QHEAD] := head;
+
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
+    result := queue![QDATA][head];
+    Unbind(queue![QDATA][head]);
+    if head <> tail then
+        if head = last then
+            head := 1;
+        else
+            head := head + 1;
+        fi;
+        queue![QHEAD] := head;
+    fi;
+
     return result;
-  fi;
-  return fail;
 end);
 
 InstallGlobalFunction(PlistQueuePopBack,
 function(queue)
-  local head, tail, last, result;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  if head <> tail then
-    if tail = 1 then
-      tail := last;
-    else
-      tail := tail - 1;
+    local head, tail, last, result;
+
+    if IsEmpty(queue) then
+        return fail;
     fi;
+
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+
     result := queue![QDATA][tail];
-    queue![QDATA][tail] := fail;
-    queue![QTAIL] := tail;
+    Unbind(queue![QDATA][tail]);
+
+    if head <> tail then
+        if tail = 1 then
+            tail := last;
+        else
+            tail := tail - 1;
+        fi;
+        queue![QTAIL] := tail;
+    fi;
+
     return result;
-  fi;
-  return fail;
+end);
+
+InstallGlobalFunction(PlistQueuePeekFront,
+function(queue)
+    if IsEmpty(queue) then
+        return fail;
+    fi;
+    return queue![QDATA][queue![QHEAD]];
+end);
+
+InstallGlobalFunction(PlistQueuePeekBack,
+function(queue)
+    if IsEmpty(queue) then
+        return fail;
+    fi;
+    return queue![QDATA][queue![QTAIL]];
 end);
 
 InstallGlobalFunction(PlistQueueExpand,
 function(queue)
-  local result, p, head, tail, last;
-  head := queue![QHEAD];
-  tail := queue![QTAIL];
-  last := queue![QCAPACITY];
-  queue![QCAPACITY] := queue![QCAPACITY] * 2;
-  p := queue![QCAPACITY];
-  result := EmptyPlist(p);
-  while p > 0 do
-    result[p] := fail;
-    p := p - 1;
-  od;
-  p := 1;
-  while head <> tail do
+    local result, p, head, tail, last, factor;
+    head := queue![QHEAD];
+    tail := queue![QTAIL];
+    last := queue![QCAPACITY];
+    factor := queue![QFACTOR];
```

Comment:
By the way, I really wonder why this is a positional object, since you effectively use it like a component object. Perhaps this is to make it easier to combine with C code?